src/pybind/mgr/cephadm/services/cephadmservice.py

```diff
@@ -984,31 +984,40 @@ class RgwService(CephService):
     def allow_colo(self) -> bool:
         return True
 
-    def config(self, spec: RGWSpec) -> None:  # type: ignore
+    def set_multisite_config(self, spec: RGWSpec) -> None:
         assert self.TYPE == spec.service_type
 
-        # set rgw_realm rgw_zonegroup and rgw_zone, if present
-        if spec.rgw_realm:
-            ret, out, err = self.mgr.check_mon_command({
-                'prefix': 'config set',
-                'who': f"{utils.name_to_config_section('rgw')}.{spec.service_id}",
-                'name': 'rgw_realm',
-                'value': spec.rgw_realm,
-            })
-        if spec.rgw_zonegroup:
-            ret, out, err = self.mgr.check_mon_command({
-                'prefix': 'config set',
+        config_keys = ['rgw_realm', 'rgw_zonegroup', 'rgw_zone']
+        for config_name in config_keys:
+            config_value = getattr(spec, config_name, None)
+            if not config_value:
+                logger.debug(f"No value for {config_name} in spec, skipping...")
+                continue
+            ret, current_value, err = self.mgr.check_mon_command({
+                'prefix': 'config get',
                 'who': f"{utils.name_to_config_section('rgw')}.{spec.service_id}",
-                'name': 'rgw_zonegroup',
-                'value': spec.rgw_zonegroup,
+                'name': config_name,
             })
-        if spec.rgw_zone:
-            ret, out, err = self.mgr.check_mon_command({
+            if ret != 0:
+                raise OrchestratorError(f"Failed to get config for {config_name}: {err}")
+            if current_value == config_value:
```

Comment:
I don't think checking the config option's value is all that much faster compared to just setting it. Is there any reason we need to do this check every time instead of just always running the set command?

Reply:
The `config` method is also called in the `finally` block of this method - https://github.com/ceph/ceph/blob/e48735c1f248082fd8fdfa0591aa56c11099ae81/src/pybind/mgr/cephadm/serve.py#L1055. So, these configs will be set twice as per the implementation. I was not sure if we should allow setting it twice or checking before hand if the values are set and avoid re-setting. I can remove it as well as per your suggestion, let me know.

Reply:
Updated the code, to juts set the config values rather than checking before hand